File: src/test/java/work/lclpnet/ap2/game/knockout/KnockoutWorldCrumbleTest.java
package work.lclpnet.ap2.game.knockout;

import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;

import static org.junit.jupiter.api.Assertions.assertEquals;

class KnockoutWorldCrumbleTest {

    private static final String[] expectedDistances = new String[] {
            """
1 1 1\s
1 0 1\s
1 1 1\s
""",
            """
3 2 2 2 3\s
2 1 1 1 2\s
2 1 0 1 2\s
2 1 1 1 2\s
3 2 2 2 3\s
""",
            """
4 4 3 3 3 4 4\s
4 3 2 2 2 3 4\s
3 2 1 1 1 2 3\s
3 2 1 0 1 2 3\s
3 2 1 1 1 2 3\s
4 3 2 2 2 3 4\s
4 4 3 3 3 4 4\s
""",
            """
6 5 4 4 4 4 4 5 6\s
5 4 4 3 3 3 4 4 5\s
4 4 3 2 2 2 3 4 4\s
4 3 2 1 1 1 2 3 4\s
4 3 2 1 0 1 2 3 4\s
4 3 2 1 1 1 2 3 4\s
4 4 3 2 2 2 3 4 4\s
5 4 4 3 3 3 4 4 5\s
6 5 4 4 4 4 4 5 6\s
"""
    };

    @ParameterizedTest
    @ValueSource(ints = {1, 2, 3, 4})
    void buildDistancesArray(int radius) {
        var distances = KnockoutWorldCrumble.buildDistancesArray(radius);

        final int len = 2 * radius + 1;

        StringBuilder builder = new StringBuilder();

        for (int z = 0; z < len; z++) {
            for (int x = 0; x < len; x++) {
                builder.append(distances[x][z]).append(' ');
            }

            builder.append('\n');
        }

        assertEquals(expectedDistances[radius - 1], builder.toString());
    }
}

File: src/minigames/knockout/src/test/kotlin/work/lclpnet/ap2/game/knockout/KnockoutWorldCrumbleTest.kt
package work.lclpnet.ap2.game.knockout

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.ValueSource

class KnockoutWorldCrumbleTest {

    @ParameterizedTest
    @ValueSource(ints = [1, 2, 3, 4])
    fun `buildDistancesArray builds correct distances`(radius: Int) {
        val distances = buildDistancesArray(radius)

        val len = 2 * radius + 1

        val builder = StringBuilder()

        for (z in 0 until len) {
            for (x in 0 until len) {
                builder.append(distances[x][z]).append(' ')
            }
            builder.append('\n')
        }

        assertEquals(expectedDistances[radius - 1], builder.toString())
    }

    companion object {
        private val expectedDistances = arrayOf(
            "1 1 1 \n1 0 1 \n1 1 1 \n",
            "3 2 2 2 3 \n2 1 1 1 2 \n2 1 0 1 2 \n2 1 1 1 2 \n3 2 2 2 3 \n",
            "4 4 3 3 3 4 4 \n4 3 2 2 2 3 4 \n3 2 1 1 1 2 3 \n3 2 1 0 1 2 3 \n3 2 1 1 1 2 3 \n4 3 2 2 2 3 4 \n4 4 3 3 3 4 4 \n",
            "6 5 4 4 4 4 4 5 6 \n5 4 4 3 3 3 4 4 5 \n4 4 3 2 2 2 3 4 4 \n4 3 2 1 1 1 2 3 4 \n4 3 2 1 0 1 2 3 4 \n4 3 2 1 1 1 2 3 4 \n4 4 3 2 2 2 3 4 4 \n5 4 4 3 3 3 4 4 5 \n6 5 4 4 4 4 4 5 6 \n"
        )
    }
}
